cloudapi/logs.go

```diff
@@ -47,6 +51,23 @@ type msgStreams struct {
 	Values [][2]string       `json:"values"` // this can be optimized
 }
 
+func (ms msgStreams) LatestTimestamp() (ts int64) {
+	if len(ms.Values) < 1 {
+		return
+	}
+	for i := 0; i < len(ms.Values); i++ {
+		raw := ms.Values[i][0]
+		unix, err := strconv.ParseInt(raw, 10, 64)
+		if err != nil {
+			return
+		}
+		if unix > ts {
+			ts = unix
+		}
+	}
+	return
+}
+
```

Comment:
Loki should always (per definition) retunr the logs in *order* so in our case if I remember correctly the last message is always *the* last message and as such with the latest timestamp

Reply:
>For any single log stream, logs must always be sent in increasing time order. If a log is received with a timestamp older than the most recent log received for that stream, that log will be dropped.

Apparently, you're right. In the case, the client (k6) is sending it out of order then loki will drop it. So we can safely consider the last item in the slice as the latest.

Reply:
Not only that but tail specifically returns them ordered as otherwise logging them will be pretty bad ;) .


 Not that we currently interleave log messages from different streams when we logged them on the screen .... as I've just noticed :facepalm: